hail/python/hailtop/batch/batch.py

```diff
@@ -207,6 +207,11 @@ def __init__(self,
 
         self._batch_handle: Optional[_bc.Batch] = None
 
+    def set_gcs_requester_pays_project(self, project: Optional[str]) -> None:
+        self.requester_pays_project = project
+        if project is not None:
+            self._backend.set_gcs_requester_pays_project(project)
```

Comment:
How come we're now providing the ability to mutate the requester pays project for a batch?

Reply:
currently, the requester pays project can be set just on the `Batch`, and there isn't an issue because in `ServiceBackend.run` the `Batch` gets passed in as an argument, and the project name is grabbed from the field on the `Batch`. since the cold storage checks need to read bucket info before the `run` method is called, they need to have the requester pays project name available, but don't have direct access to the `Batch` object. i originally tried passing the project name to `validate_file`, but ultimately it needs to be set on the `GoogleStorageClient`, so i was just setting that field from `validate_file`, which seems non-ideal, so instead i took the approach of setting the project name on the storage client as soon as it gets set on the `Batch`.

Reply:
Ah, ok, thanks for the explanation, this makes sense. I think there's an unfortunate side-effect here though (in the literal PL sense). Mutating the `_gcs_requester_pays_configuration` in `GoogleStorageClient` is not safe if there are any in-flight requests happening in other `asyncio` tasks. This is maybe OK, since we are expecting such mutation to only come from synchronous code running outside an event loop, but I think if we have to do this we should at least leave a big comment somewhere.

Aside from that, it feels incorrect for the `Batch` to mutate the requester pays project on the `ServiceBackend`, since the `ServiceBackend` and its FS can be reused to create multiple batches. For example, this feels incorrect:

```python
backend = hb.ServiceBackend(gcs_requester_pays_configuration='project-x')
b = hb.Batch(backend, gcs_requester_pays_project='project-y')
… 
b2 = hb.Batch(backend)  # Wouldn't expect to use 'project-y'
```

Reply:
Looking at the implementation on `main` a little further raised a question: Is `batch.requester_pays_project` intended to be used on the user's laptop? It has not been used up until now, and in `main` that field is only used when adding it to the batch spec. Similarly, I don't see anywhere where `ServiceBackend.gcs_requester_pays_configuration` makes it into the batch spec and is propagated to jobs. As it stands right now, it *looks* like `Servicebackend.gcs_requester_pays_configuration` is intended for client-side use while `Batch.requester_pays_project` is for configuring jobs. I have no idea whether this is accurate or as intended. What functionality *should* these fields have?

Reply:
good point; the thing that would make the most sense to me would be that the value on the `Backend` would be the default for `Batch`es created using it, and the arg in the `Batch` constructor would override it if provided, in which case it would actually be correct to pass the project name from the `Batch` through as an arg to `validate_file`, and it should similarly be passed along when creating `Job`s on the `Batch` or calling `run` on any `Backend`. but i would also expect there to be something like

```python
self.requester_pays_project = requester_pays_project if requester_pays_project is not None else getattr(backend._fs._gcs_kwargs, "gcs_requester_pays_configuration", None)
```

in `Batch.__init__`. i'll ask for clarification on zulip and then circle back here

Reply:
Fully agree with this take